Introduce STM32MP25x platforms minimal support

GseoC opened the pull request:
This P-R aims to bring minimal platform support for STM32MP25x 64 bits platforms in OP-TEE.
You'll find a resume of the features in the first commit on this series.

This P-R cleans some code in GPIO and UART driver as well that blocked the compilation of another platform using STM32 drivers.

A weak platform-specific plat_panic() is added to add platform-specific behavior in case of panic().

@etienne-lms As discussed in private, I put you as maintainer of the platform as well as myself.

@etienne-lms @tbourgoi, can you please review this P-R?

GseoC commented:
@jforissier, any idea why the Qemu check failed to clone some sub modules?

jforissier commented:
> @jforissier, any idea why the Qemu check failed to clone some sub modules?

GitLab is having problems: https://status.gitlab.com/. I will restart the jobs when things are back to normal.

tbourgoi commented on core/arch/arm/dts/stm32mp25-pinctrl.dtsi:
```diff
@@ -0,0 +1,7 @@
+// SPDX-License-Identifier: (GPL-2.0-or-later OR BSD-3-Clause)
+/*
+ * Copyright (C) STMicroelectronics 2021 - All Rights Reserved
```
Could you update the year please : 2023

tbourgoi commented on core/arch/arm/dts/stm32mp25xc.dtsi:
```diff
@@ -0,0 +1,8 @@
+// SPDX-License-Identifier: (GPL-2.0-or-later OR BSD-3-Clause)
+/*
+ * Copyright (C) STMicroelectronics 2021 - All Rights Reserved
```
Year also here

tbourgoi commented on core/arch/arm/plat-stm32mp2/conf.mk:
```diff
@@ -0,0 +1,74 @@
+endif
+CFG_EMBED_DTB_SOURCE_FILE ?= $(flavor_dts_file-$(PLATFORM_FLAVOR))
+endif
+CFG_EMBED_DTB_SOURCE_FILE ?= stm32mp257f-ev1.dts
```
I think using `flavor_dts_file-257F_EV` is more appropriate here (in case the dts filename changes

GseoC replied:
If the name changes, I'd still have to change the name of the flavorlist so I think it's ok

tbourgoi commented on core/arch/arm/plat-stm32mp2/main.c:
```diff
@@ -0,0 +1,223 @@
+		paddr_t pa;
+		bool secure;
+	} uarts[] = {
+		[0] = { .pa = 0 },
```
nitpicking, aligning `.secure` might be easier to read

GseoC replied:
I think it's ok as it is

GseoC commented on core/arch/arm/plat-stm32mp2/conf.mk:
```diff
@@ -0,0 +1,74 @@
+flavor_dts_file-257F_EV = stm32mp257f-ev1.dts
```
EV1

GseoC commented on core/arch/arm/dts/stm32mp25xf.dtsi:
```diff
@@ -0,0 +1,8 @@
+// SPDX-License-Identifier: (GPL-2.0-or-later OR BSD-3-Clause)
+/*
+ * Copyright (C) STMicroelectronics 2021 - All Rights Reserved
```
Here as well

GseoC commented on core/arch/arm/dts/stm32mp25xxak-pinctrl.dtsi:
```diff
@@ -0,0 +1,69 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR BSD-3-Clause)
+/*
+ * Copyright (C) STMicroelectronics 2021 - All Rights Reserved
```
Date to check in all files

tbourgoi commented on core/arch/arm/plat-stm32mp2/platform_config.h:
```diff
@@ -0,0 +1,115 @@
+
+/* SoC interface registers base address */
+#define I2C4_BASE			0x40150000
+#define I2C6_BASE			0x40170000
```
These two defines are not at the right place based on their addresses.
They should be between UART5 and UART6

etienne-lms replied:
not yet addressed

GseoC replied:
Thanks for noticing, I'll change that right away

etienne-lms commented on core/arch/arm/plat-stm32mp2/main.c:
```diff
@@ -163,3 +163,53 @@ void may_spin_unlock(unsigned int *lock, uint32_t exceptions)
+	if (CFG_TEE_CORE_NB_CORE == 1)
+		return false;
+
+	return true;
```
`return CFG_TEE_CORE_NB_CORE == 1;`

etienne-lms commented on core/arch/arm/plat-stm32mp2/main.c:
```diff
@@ -163,3 +163,53 @@ void may_spin_unlock(unsigned int *lock, uint32_t exceptions)
+	return true;
+}
+
+/* SGI9 (secure SGI 1) informs targeted CPU it shall reset */
```
"it shall ~~reset~~ **abort execution**". 
We expect reset will come from a watchdog, as currently implemented in generic `panici()` funtion.

etienne-lms commented on core/arch/arm/plat-stm32mp2/main.c:
```diff
@@ -207,8 +209,14 @@ void __noreturn plat_panic(void)
+	if (res)
+		panic();
+
+	interrupt_enable(sgi9_reset_handler.chip, sgi9_reset_handler.it);
```
For commit "core: panic: platform specific panic sequence": 
These changes in _core/arch/arm/plat-stm32mp2/main.c_ should be squashed in commit "plat-stm32mp2: implement plat_panic() function" where they belong.

GseoC replied:
Will remove this commit from this series

etienne-lms commented on core/arch/arm/plat-stm32mp2/conf.mk:
```diff
@@ -0,0 +1,74 @@
+# Uncomment the line below to only support 64bit TA
+supported-ta-targets ?= ta_arm64
+
+$(call force,CFG_ARM_GIC_PM,y)
```
remove this line. This relates to an [old P-R](https://github.com/OP-TEE/optee_os/pull/2709) that did not complete its journey. Maybe time to dig it up.

GseoC replied:
Sure :)

etienne-lms commented on core/arch/arm/plat-stm32mp2/conf.mk:
```diff
@@ -0,0 +1,74 @@
+$(call force,CFG_PM,y)
+$(call force,CFG_SECURE_TIME_SOURCE_CNTPCT,y)
+$(call force,CFG_DT,y)
+$(call force,CFG_GENERIC_BOOT,y)
```
can remove undefined `CFG_PM` and `CFG_GENERIC_BOOT`

etienne-lms commented on core/arch/arm/plat-stm32mp2/conf.mk:
```diff
@@ -0,0 +1,74 @@
+$(call force,CFG_DRIVERS_CLK_DT,y)
+$(call force,CFG_DRIVERS_CLK_EARLY_PROBE,y)
+$(call force,CFG_TEE_CORE_EMBED_INTERNAL_TESTS,n)
+
```
what about sorting this `$(call force,...)` config list (lines 29 to 39) in alphabetical order?
and also move lines 21, 22 and 25 into this forced config list.

etienne-lms reviewed (commented):
`Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>`. for commits
"drivers: stm32_gpio: remove unused APIs",
"dts: stm32: introduce STM32MP25 SoCs family",
"dts: stm32: add stm32mp257f-ev1 board support",
"MAINTAINERS: update STMicroelectronics platforms and drivers entries" and 
"ci: azure: add STM32MP2x platforms default build".

For commit "core: panic: platform specific panic sequence", i think this change deserves a dedicated P-R so that @jenswi-linaro, @jforissier and other people can have a specific look at it. It's a bit hidden inside an ST platforms specific series. @GseoC, would you mind creating a P-R for that and propose a stm32mp1 port? (if you're too busy, i'll do it, based on your change).

jforissier commented:
"ci: azure: add STM32MP2x platforms default build": please drop "azure:". CI is now using GitHub actions which may be based on Azure but that's not visible.

GseoC commented:
> `Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>`. for commits "drivers: stm32_gpio: remove unused APIs", "dts: stm32: introduce STM32MP25 SoCs family", "dts: stm32: add stm32mp257f-ev1 board support", "MAINTAINERS: update STMicroelectronics platforms and drivers entries" and "ci: azure: add STM32MP2x platforms default build".
> 
> For commit "core: panic: platform specific panic sequence", i think this change deserves a dedicated P-R so that @jenswi-linaro, @jforissier and other people can have a specific look at it. It's a bit hidden inside an ST platforms specific series. @GseoC, would you mind creating a P-R for that and propose a stm32mp1 port? (if you're too busy, i'll do it, based on your change).

Hi thanks for the review. I don't mind creating a specific P-R for this commit :)

GseoC commented on core/arch/arm/plat-stm32mp2/conf.mk:
```diff
@@ -0,0 +1,74 @@
+$(call force,CFG_DRIVERS_CLK,y)
+$(call force,CFG_DRIVERS_CLK_DT,y)
+$(call force,CFG_DRIVERS_CLK_EARLY_PROBE,y)
+$(call force,CFG_TEE_CORE_EMBED_INTERNAL_TESTS,n)
```
Agree, I'll remove this line

GseoC commented on core/arch/arm/plat-stm32mp2/conf.mk:
```diff
@@ -0,0 +1,74 @@
+# arm-v8 platforms
+include core/arch/arm/cpu/cortex-armv8-0.mk
+$(call force,CFG_WITH_ARM_TRUSTED_FW,y)
+# Uncomment the line below to only support 64bit TA
```
Do you mean the comment and move the include at the top of the conf.mk?

etienne-lms replied:
I think both inline comments `# arm_v8 ...` and `Uncomment ...` as useless.
`include core/arch/arm/cpu/cortex-armv8-0.mk` could be placed above line 21 (before the `$call force, ...)` series.

GseoC commented on core/arch/arm/plat-stm32mp2/conf.mk:
```diff
@@ -0,0 +1,74 @@
+CFG_STM32_GPIO ?= y
+ifeq ($(CFG_STM32_GPIO),y)
+$(call force,CFG_DRIVERS_GPIO,y)
+$(call force,CFG_DRIVERS_PINCTRL,y)
```
Ok, makes sense as the clk driver is forced to y

GseoC replied:
I'll sort the config as well

etienne-lms commented on core/arch/arm/plat-stm32mp2/main.c:
```diff
@@ -0,0 +1,223 @@
+	console_data.clock = NULL;
+
+	console_data.secure = uarts[CFG_STM32_EARLY_CONSOLE_UART].secure;
+#if CFG_STM32_UART
```
If `CFG_STM32_UART` is disabled, `register_serial_console()` should not be called unless what `console_putc()`/`console_flush()` will segfault.

GseoC commented:
> "ci: azure: add STM32MP2x platforms default build": please drop "azure:". CI is now using GitHub actions which may be based on Azure but that's not visible.

Ok, I will remove it

GseoC commented on core/arch/arm/plat-stm32mp2/stm32_util.h:
```diff
@@ -0,0 +1,38 @@
+ * lock is NULL.
+ */
+uint32_t may_spin_lock(unsigned int *lock);
+void may_spin_unlock(unsigned int *lock, uint32_t exceptions);
```
Will move to a macro definition to standard API

GseoC commented:
- Applied tags
- Squashed drivers: stm32_uart: support STM32MP25x platforms in plat-stm32mp2: add new platform support to ensure compilation at all stages
- Remove plat_panic() related patches from this P-R
- Made the fixes
- Removed azur: from ci: azure: add STM32MP2x platforms default build

etienne-lms commented on core/arch/arm/plat-stm32mp2/stm32_util.h:
```diff
@@ -33,6 +33,7 @@ static inline void stm32mp_register_gpioz_pin_count(size_t count __unused) { }
  * Generic spinlock function that bypass spinlock if MMU is disabled or
  * lock is NULL.
  */
```
can remove the above inline comment.
Actually I think you can remove all the above inline comments.

GseoC replied:
Removed all inline comments

etienne-lms commented on core/arch/arm/plat-stm32mp2/platform_config.h:
```diff
@@ -0,0 +1,112 @@
+#define FMC_BASE			0x48200000
+#define PCIE_BASE			0x48400000
+#define A35SSC_BASE			0x48800000
+#define GIC_BASE			0x4ac00000ul
```
`ul` not needed

etienne-lms commented on core/arch/arm/plat-stm32mp2/platform_config.h:
```diff
@@ -0,0 +1,112 @@
+#define PCIE_BASE			0x48400000
+#define A35SSC_BASE			0x48800000
+#define GIC_BASE			0x4ac00000ul
+#define DDR_BASE			0x80000000ul
```
use `UL()`

etienne-lms commented on core/arch/arm/plat-stm32mp2/platform_config.h:
```diff
@@ -0,0 +1,112 @@
+#define GIC_SPI_UART4			126
+
+#define CONSOLE_UART_BASE		STM32MP1_DEBUG_USART_BASE
+#define CONSOLE_UART_SIZE		1024
```
can remove `CONSOLE_UART_BASE`, `CONSOLE_UART_SIZE` and `STM32MP1_DEBUG_USART_BASE`. Unused.

etienne-lms reviewed (commented):
Few more comments. The whole looks ok once those are addressed. I'm fine I squash fixup commits for a final review round.

GseoC commented:
Comments addressed and review patches squashed

jforissier commented:
A sentence is repeated in the commit description for "plat-stm32mp2: add new platform support". Also s/gic/GIC/.

GseoC commented:
> A sentence is repeated in the commit description for "plat-stm32mp2: add new platform support". Also s/gic/GIC/.

Oops, thank you.

Commit message updated

etienne-lms commented on core/arch/arm/plat-stm32mp2/conf.mk:
```diff
@@ -0,0 +1,62 @@
+$(call force,CFG_STM32MP25,y)
+endif
+
+ifeq ($(filter $(CFG_STM32MP25),y),)
```
`ifneq ($(CFG_STM32MP25),y)`

etienne-lms commented on core/arch/arm/plat-stm32mp2/conf.mk:
```diff
@@ -0,0 +1,62 @@
+$(call force,CFG_ARM64_core,y)
+$(call force,CFG_WITH_ARM_TRUSTED_FW,y)
+$(call force,CFG_WITH_LPAE,y)
+supported-ta-targets ?= ta_arm64
```
I suggest to move the above line  out of this call-force series and reorder to 3 config setting above into the below series:
```patch
 include core/arch/arm/cpu/cortex-armv8-0.mk
 supported-ta-targets ?= ta_arm64

+$(call force,CFG_ARM64_core,y)
 $(call force,CFG_DRIVERS_CLK,y)
 ...
 $(call force,CFG_SECURE_TIME_SOURCE_CNTPCT,y)
+$(call force,CFG_WITH_ARM_TRUSTED_FW,y)
+$(call force,CFG_WITH_LPAE,y)

etienne-lms commented on core/arch/arm/plat-stm32mp2/conf.mk:
```diff
@@ -0,0 +1,62 @@
+CFG_TEE_CORE_NB_CORE ?= 2
+
+CFG_STM32_GPIO ?= y
+CFG_STM32_SHARED_IO ?= y
```
I think this switch is required. `$(call force,CFG_STM32_SHARED_IO,y)`

GseoC replied:
Actually it is not, let me remove this switch.

etienne-lms commented on core/arch/arm/plat-stm32mp2/stm32_util.h:
```diff
@@ -0,0 +1,26 @@
+#include <assert.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <types_ext.h>
```
`#include <kernel/spinlock.>`
and remove inclusions of _assert.h_, _stdbool.h_

GseoC replied:
Ok but I'll add a patch on top of this series to add missing assert.h include in stm32_opp driver.

Seen by trying to remove 
#include <assert.h>
#include <stdbool.h>

and not adding:
#include <kernel/spinlock.h>

As assert.h is included in #include <kernel/spinlock.h>, it's not seen otherwise.

etienne-lms commented on core/drivers/stm32_uart.c:
```diff
@@ -113,19 +114,32 @@ static void register_secure_uart(struct stm32_uart_pdata *pd)
 	size_t __maybe_unused n = 0;
 
 	stm32mp_register_secure_periph_iomem(pd->base.pa);
+#ifndef CFG_STM32MP25
```
for consistency, could you condition also the above line with `#ifndef CFG_STM32MP25`
(and by the way, remove this useless leftover `n` local variable, if you don't mind)

GseoC commented:
Addressed comments

etienne-lms commented on core/arch/arm/plat-stm32mp2/conf.mk:
```diff
@@ -0,0 +1,61 @@
+
+# Default enable some test facitilites
+CFG_TEE_CORE_EMBED_INTERNAL_TESTS ?= y
+CFG_WITH_STATS ?= y
```
nitpicking: should remove `CFG_TEE_CORE_EMBED_INTERNAL_TESTS ?= y`. There is a default config in _mk/config.mk_ that fulfills this requirement IMHO.

etienne-lms commented on core/arch/arm/plat-stm32mp2/platform_config.h:
```diff
@@ -0,0 +1,108 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) 2021, STMicroelectronics
```
2023 :)

etienne-lms commented on core/arch/arm/plat-stm32mp2/main.c:
```diff
@@ -0,0 +1,143 @@
+
+	return TEE_SUCCESS;
+}
+service_init(platform_banner);
```
add an empty line above.

etienne-lms commented on core/arch/arm/plat-stm32mp2/conf.mk:
```diff
@@ -0,0 +1,61 @@
+$(call force,CFG_ARM64_core,y)
+$(call force,CFG_DRIVERS_CLK,y)
+$(call force,CFG_DRIVERS_CLK_DT,y)
+$(call force,CFG_DRIVERS_CLK_EARLY_PROBE,y)
```
Do we really need this? I don't think so.

GseoC commented:
Comments addressed

etienne-lms reviewed (commented):
Thanks for your patience.
`Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commit
"plat-stm32mp2: add new platform support".

GseoC commented:
> Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>

No problem, we're better off with a sane base.

Tag applied.
